### PR TITLE
Stop loop in BitTrie.addPrefixRange if length exceeds max

### PR DESCRIPTION
**Pull request by corinaminer**
Fixes first part of #653. Next step is adding tests.

---
This change is [Reviewable](https://reviewable.io/reviews/batfish/batfish/671)

### CODE REVIEW

File: projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java [hunk 1]
```diff
@@ -798,8 +798,7 @@ public void initTestrig(
     // Create metadata file (RELPATH_METADATA_FILE is "metadata.json")
     Path metadataPath = testrigDir.resolve(BfConsts.RELPATH_METADATA_FILE);
     long time = (new java.util.Date()).getTime();
-    CommonUtil.writeFile(metadataPath, "{\n\t\"timestamp\": "+Long.toString(time)+"\n}");
-
+    CommonUtil.writeFile(metadataPath, "{\n\t\"timestamp\": " + Long.toString(time) + "\n}");
```
Comment (corinaminer) [inline projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java]: Don't know why I didn't have to reformat this file before merging the PR that introduced it, but travis wasn't having any of it this time

**Comment by dhalperi** [comment]
*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 801 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz-wQbOFomEphwEIkMb-r1-801:-Kz02OuXNhTq58c9xrQU:b-dbqoft) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L801)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Don't know why I didn't have to reformat this file before merging the PR that introduced this change, but travis wasn't having any of it this time
</blockquote></details>

Because Travis wasn't green, but your PR was merged anyway.

https://travis-ci.org/batfish/batfish/builds/302185113?utm_source=github_status&utm_medium=notification

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 15 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0OqIbdxSeL0VA0cbq:-Kz0OqIbdxSeL0VA0cbr:bqz93iy) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L15)):*
> ```Java
> import java.util.concurrent.ConcurrentMap;
> import javax.annotation.Nullable;
> ```

Now that you understand this class fully, would you mind adding class-level Javadoc to pay it forward? :)

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 35 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0MXS6Gtt4iPL0F1gW:-Kz0MXS6Gtt4iPL0F1gX:bmplqk3) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L35)):*
> ```Java
>       int minLength = prefixRange.getLengthRange().getStart();
>       int maxLength = prefixRange.getLengthRange().getEnd();
>       int matchLength = prefixRange.getPrefix().getPrefixLength();
> ```

This part looks like it could be simplified, especially if enhanced by comments:

```java
// The minimum length of the range may be shorter than the actual prefix length.
// If so, we need to specially handle all shorter prefixes with a custom address and bitset.
int minLength = prefixRange.getLengthRange().getStart();
int maxLength = Math.min(prefixRange.getLengthRange().getEnd(), prefixLength - 1);
for (int i = minLength; i <= maxLength; ++i) {
   ...
}

// Add the prefix-range as-is
BitSet bits = ...
add.
```

(also, note that `matchLength` appears to be `prefixLength`, just recomputed in 1 line instead of two).

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 44 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0MK-gwLvwC5L94s27:-Kz0MK-gwLvwC5L94s28:b-ej6ors) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L44)):*
> ```Java
>         _root.addPrefixRange(currentPrefixRange, currentBits, currentLength, 0);
>       }
>       _root.addPrefixRange(prefixRange, bits, prefixLength, 0);
> ```

agree on the need for tests :)

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 213 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz02aDkcowdzsEqBLfj:-Kz02aDkcowdzsEqBLfk:bxy1xql) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L213)):*
> ```Java
> 
>   /**
>    * Adds the given prefix range into the {@code BitTrie} representing this {@code PrefixSpace}.
> ```

Javadoc comments should only address the public surface of the class. In this case, since `BitTrie` is a `private static class`, it's not part of the public surface; it's an implementation detail. (I.e., don't talk about `BitTrie` here.)

Also, `@link` instead of `@code` the class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
@corinaminer I believe this is pending on you

**Comment by corinaminer** [comment]
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 15 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0OqIbdxSeL0VA0cbq:-L-hJRb8yHUoy7rm14ry:b-eg7i0t) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L15)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Now that you understand this class fully, would you mind adding class-level Javadoc to pay it forward? :)
</blockquote></details>

"understand this class fully" is a strong term... I can put down a high-level description like "A set of Prefixes that specifies some set of IP addresses", but not sure what else I would put in class-level javadoc.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 35 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0MXS6Gtt4iPL0F1gW:-L-honZcAomsws-K1LRd:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L35)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

This part looks like it could be simplified, especially if enhanced by comments:

```java
// The minimum length of the range may be shorter than the actual prefix length.
// If so, we need to specially handle all shorter prefixes with a custom address and bitset.
int minLength = prefixRange.getLengthRange().getStart();
int maxLength = Math.min(prefixRange.getLengthRange().getEnd(), prefixLength - 1);
for (int i = minLength; i <= maxLength; ++i) {
   ...
}

// Add the prefix-range as-is
BitSet bits = ...
add.
```

(also, note that `matchLength` appears to be `prefixLength`, just recomputed in 1 line instead of two).
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 213 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz02aDkcowdzsEqBLfj:-L-hoqkvWGLimdrC9q_I:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L213)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Javadoc comments should only address the public surface of the class. In this case, since `BitTrie` is a `private static class`, it's not part of the public surface; it's an implementation detail. (I.e., don't talk about `BitTrie` here.)

Also, `@link` instead of `@code` the class.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by arifogel** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 803 at r2](https://reviewable.io:443/reviews/batfish/batfish/671#-L-hpBBYv4r1DUwiXx1U:-L-hpBBZ5PI3HjSnSLXe:b-7rcx9v) ([raw file](https://github.com/batfish/batfish/blob/16d76ccacba4b41724aa1a06256ca3aab89b074e/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L803)):*
> ```Java
>     CommonUtil.writeFile(metadataPath, "{\n\t\"timestamp\": "+Long.toString(time)+"\n}");
> ```

Removing whitespace here seems out of scope, but probably necessary to satisfy formatter check. Does this issue go away with a rebase on master?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by arifogel** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 15 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0OqIbdxSeL0VA0cbq:-L-jg6e9q80HpP9Ib6yl:b-swmdk8) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L15)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

"understand this class fully" is a strong term... I can put down a high-level description like "A set of Prefixes that specifies some set of IP addresses", but not sure what else I would put in class-level javadoc.
</blockquote></details>

Please make an effort :) Feel free to run it by Ari / me.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 35 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0MXS6Gtt4iPL0F1gW:-L-jfVNj-872PBPYgdoJ:b-h4lvtd) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L35)):*
> <= maxLength

There's an off-by-one error here somewhere. We're handling the case of `length == prefix.getPrefixLength()` twice, once in the for loop and once after it.

There are a few subtle differences in the snippet I suggested and the one you wrote, bug I no longer recall what's going on here. Hard for me to reason about right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by corinaminer** [comment]
*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java, line 35 at r1](https://reviewable.io:443/reviews/batfish/batfish/671#-Kz0MXS6Gtt4iPL0F1gW:-L-mEZD5xBfdMPWAzJM0:b-j59ykp) ([raw file](https://github.com/batfish/batfish/blob/c9453c52e9a3bcc7a3b8d957dfd0ed053604e15e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpace.java#L35)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> <= maxLength

There's an off-by-one error here somewhere. We're handling the case of `length == prefix.getPrefixLength()` twice, once in the for loop and once after it.

There are a few subtle differences in the snippet I suggested and the one you wrote, bug I no longer recall what's going on here. Hard for me to reason about right now.
</blockquote></details>

I forgot to subtract 1 from `prefixLength` at the end of your fourth line there. I think that's the only difference. Will fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by corinaminer** [comment]
@arifogel @dhalperi I added a pile of tests for `PrefixSpace`. Please check that the expected results defined in the tests are correct, and let me know what other edge cases should be tested.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671#-:-L-r_k1iyOjLYyO7NGra:b-1pznsw)*

**Comment by arifogel** [comment]
I think this is fine, modulo offline conversation about canonicalization. I will leave it to @dhalperi to give final approval.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 44 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-rajuKiIEOE4h4lK65:-L-rajuLmX3mdsyzN6u5:b-am6okf) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L44)):*
> ```Java
>     PrefixRange range = new PrefixRange("100.0.0.0/32");
>     ranges.add(range);
>     _ps = new PrefixSpace(ranges);
> ```

Optional style change:
_ps = new PrefixSpace(ImmutableSet.of(new PrefixRange("100.0.0.0/32")))

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 67 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-rbIvYM5jNbyMp2nU_:-L-rbIvYM5jNbyMp2nUa:buollnr) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L67)):*
> ```Java
>         "Shorter prefixes not included",
>         _ps.containsPrefix(new Prefix("10.10.10.0/15")),
>         equalTo(false));
> ```

Please file issue to use custom Matcher

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671#-:-L-rgpTqUtEGybOI3k2a:b-ofxj2f)*

**Comment by dhalperi** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 84 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-raPGDwxjSQPVgZyFw:-L-raPGDwxjSQPVgZyFx:bmykkzc) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L84)):*
> ```Java
>         _ps.containsPrefix(new Prefix("10.10.11.0/24")),
>         equalTo(false));
>     assertThat(
> ```

These are great tests. A few suggestions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 87 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-r_yj7tW3taUrcGkJv:-L-r_yj7tW3taUrcGkJw:bq62hgf) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L87)):*
> ```Java
>         "Prefixes with mismatch in unmasked bits included",
>         _ps.containsPrefix(new Prefix("10.10.11.0/16")),
>         equalTo(true));
> ```

check expected size too?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 87 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-ra5W4WSpA6o5h5Jw_:-L-ra5W4WSpA6o5h5Jwa:b-s45kkj) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L87)):*
>   assertThat(
>         "Prefixes with mismatch in unmasked bits included",
>         _ps.containsPrefix(new Prefix("10.10.11.0/16")),
>         equalTo(true));

check closer to the boundary conditions? E.g., I'd test `/23` and `/17` to guarantee no off-by-1 error on either side.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671#-:-L-s1H3ALrEBKLDp-OaC:bnfp4nl)*

**Comment by corinaminer** [comment]
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 87 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-r_yj7tW3taUrcGkJv:-L-sFAVrV3nyZE0ibLgm:bk0ipo5) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L87)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

check expected size too?
</blockquote></details>

Size of what? I could check the number of PrefixRanges the PrefixSpace has, but that's more thoroughly covered in `getPrefixRangesTest`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java, line 87 at r4](https://reviewable.io:443/reviews/batfish/batfish/671#-L-r_yj7tW3taUrcGkJv:-L-sVMdmr9Vif43b8FzF:b-t77xc3) ([raw file](https://github.com/batfish/batfish/blob/170b8db740e5e3793103042ed9ccf908e8a1f921/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/PrefixSpaceTest.java#L87)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Size of what? I could check the number of PrefixRanges the PrefixSpace has, but that's more thoroughly covered in `getPrefixRangesTest`.
</blockquote></details>

agreed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*

**Comment by dhalperi** [comment]
Reviewed 1 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/671)*
